### PR TITLE
Es test

### LINKED ISSUE #399: Run tests on es modules
When using `--format es` on the command line, with a grammar that includes `import` statements, I would like to be able to use `-t` and `-T`.

Import paths should be relative to either the output file (if specified), or to the grammar file (if no output specified).  Tests should run in a vm/context like they do for commonjs.

### PR DESCRIPTION
**Pull request by hildjj**
Allow testing of es modules.  Fixes #399

### CODE REVIEW

**Comment by frostburn** [comment]
Tests succeed on my end.
I'm not familiar with the issues in this PR. Is there something that I could test manually that is broken on `main` and this one fixes?

**Comment by hildjj** [comment]
This should fail on main, but pass on this branch:

```sh
echo "foo = '1'" | bin/peggy.js --format es -t 1
```

unless you run this branch on node 20.7 or earlier, then it should fail with a version error.

**Comment by hildjj** [comment]
All of the interesting stuff is now in `bin/fromMem.js`.  It shouldn't actually be security-relevant, even though it looks like it.  All of those gyrations are there so that `import` and `require` work relative to the output directory (which is the input directory by default).

**Comment by frostburn** [comment]
> This should fail on main, but pass on this branch:
> 
> ```shell
> echo "foo = '1'" | bin/peggy.js --format es -t 1
> ```
> 
> unless you run this branch on node 20.7 or earlier, then it should fail with a version error.

Apparently my version of Ubuntu ships with node version 18.13.0 so I get the expected failure. (Don't feel like messing around with updates right now...)

On `main` I get:
```bash
Error running test
Unexpected token 'export'
```

**Comment by hildjj** [comment]
Good, that works as expected.  I like to use `npm -g install nve` and `nve 21 ./bin/peggy ...` to use different node versions without having to change what is installed globally.

**Comment by frostburn** [comment]
Thanks!
```bash
$ echo "foo = '1'" | nve 21 node --experimental-vm-modules ./bin/peggy.js --format es -t 1
'1'
```

**Review by hildjj — COMMENTED** [review]
Check a few things.

File: bin/fromMem.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    initializeImportMeta(meta) {
+      meta.url = String(url.pathToFileURL(options.filename));
+    },
+    importModuleDynamically(specifier) {
+      return import(resolveIfNeeded(dirname, specifier));
+    },
+  });
+
+  await mod.link(async(specifier, referencingModule) => {
+    const resolvedSpecifier = resolveIfNeeded(dirname, specifier);
```
[the remark this answers is not in the document]
Comment (hildjj) [reply bin/fromMem.js:125]: This seems to work fine in practice.  Shrug, wait for the bug report.

File: bin/fromMem.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    const targetModule = await import(resolvedSpecifier);
+    const exports = Object.keys(targetModule);
+
+    // DO NOT change function to () =>, or `this` will be wrong.
+    return new vm.SyntheticModule(exports, function() {
+      for (const e of exports) {
+        this.setExport(e, targetModule[e]);
+      }
+    }, {
```
[the remark this answers is not in the document]
Comment (hildjj) [reply bin/fromMem.js:134]: Seems to work fine in practice.

**Pull request merged**